File: samples/ScheduledTasks.java
// Spring @Scheduled. Open in the sandbox IDE; the expected result is above each expression.
//
// Nothing here needs Spring on the classpath - the annotator matches @Scheduled by its simple
// name, so an unresolved annotation still gets read.
package samples;

import org.springframework.scheduling.annotation.Scheduled;

public class ScheduledTasks {

    // OK - "At 12:00 AM". Six fields, seconds first: the same text is a hard error in a
    // workflow file, and the plugin reads it correctly in both places.
    @Scheduled(cron = "0 0 0 * * ?")
    public void midnight() {}

    // OK - "Every 30 seconds"
    @Scheduled(cron = "*/30 * * * * *")
    public void heartbeat() {}

    // OK - "At 09:15 AM, Monday through Friday"
    @Scheduled(cron = "0 15 9 * * MON-FRI")
    public void weekdayOpen() {}

    // WARNING (day OR) - the 1st AND every Monday, not only Mondays that are the 1st
    @Scheduled(cron = "0 0 0 1 * MON")
    public void billing() {}

    // WARNING (uneven step) - 0, 7 ... 56, then 0 again, a 4-minute gap across the hour
    @Scheduled(cron = "0 */7 * * * *")
    public void uneven() {}

    // ERROR on the seconds field only - the squiggle should cover `60`, not the whole string
    @Scheduled(cron = "60 0 0 * * *")
    public void badSecond() {}

    // ERROR - five fields is Unix cron; @Scheduled always wants six
    @Scheduled(cron = "0 0 * * *")
    public void tooFewFields() {}

    // Silent - a property placeholder is not an expression to explain
    @Scheduled(cron = "${report.schedule}")
    public void externalised() {}

    // Silent - fixedDelay is milliseconds, and the annotator must not touch other attributes
    @Scheduled(fixedDelayString = "0 0 0 * * ?")
    public void notACronAttribute() {}
}

File: samples/ScheduledTasks.kt
// Spring @Scheduled in Kotlin. Same annotation, different PSI - a Kotlin string is a template
// built from entries rather than a literal, which is why this needs its own recogniser.
package samples

import org.springframework.scheduling.annotation.Scheduled

class ScheduledTasks {

    // OK - "At 12:00 AM". Six fields, seconds first: the same text is a hard error in a
    // workflow file, and the plugin reads it correctly in both places.
    @Scheduled(cron = "0 0 0 * * ?")
    fun midnight() {}

    // OK - "Every 30 seconds"
    @Scheduled(cron = "*/30 * * * * *")
    fun heartbeat() {}

    // WARNING (day OR) - the 1st AND every Monday, not only Mondays that are the 1st
    @Scheduled(cron = "0 0 0 1 * MON")
    fun billing() {}

    // ERROR on the seconds field only - the squiggle should cover `60`
    @Scheduled(cron = "60 0 0 * * *")
    fun badSecond() {}

    // ERROR - five fields is Unix cron; @Scheduled always wants six
    @Scheduled(cron = "0 0 * * *")
    fun tooFewFields() {}

    // Silent - an escaped placeholder is not an expression to explain
    @Scheduled(cron = "\${report.schedule}")
    fun externalised() {}

    // Silent - a template with an interpolation is not knowable without running the program
    @Scheduled(cron = "0 0 $HOUR * * *")
    fun interpolated() {}

    // Silent - fixedDelayString is milliseconds, not a schedule
    @Scheduled(fixedDelayString = "0 0 0 * * ?")
    fun notACronAttribute() {}

    private companion object {
        const val HOUR = "3"
    }
}
